// nem2-sdk/dist/src/infrastructure/builders/AccountRestrictionsMosaicTransaction.js
"use strict";
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @module transactions/AccountRestrictionsMosaicTransaction
 */
const TransactionType_1 = require("../../model/transaction/TransactionType");
const AccountRestrictionsMosaicTransactionBuffer_1 = require("../buffers/AccountRestrictionsMosaicTransactionBuffer");
const AccountRestrictionsMosaicModificationTransactionSchema_1 = require("../schemas/AccountRestrictionsMosaicModificationTransactionSchema");
const VerifiableTransaction_1 = require("./VerifiableTransaction");
const { AccountRestrictionsMosaicTransactionBuffer, RestrictionMosaicModificationBuffer, } = AccountRestrictionsMosaicTransactionBuffer_1.default.Buffers;
const flatbuffers_1 = require("flatbuffers");
class AccountRestrictionsMosaicTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes) {
        super(bytes, AccountRestrictionsMosaicModificationTransactionSchema_1.default);
    }
}
exports.default = AccountRestrictionsMosaicTransaction;
// tslint:disable-next-line:max-classes-per-file
class Builder {
    constructor() {
        this.maxFee = [0, 0];
        this.type = TransactionType_1.TransactionType.MODIFY_ACCOUNT_RESTRICTION_MOSAIC;
    }
    addFee(maxFee) {
        this.maxFee = maxFee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addRestrictionType(restrictionType) {
        this.restrictionType = restrictionType;
        return this;
    }
    addModifications(modifications) {
        this.modifications = modifications;
        return this;
    }
    build() {
        const builder = new flatbuffers_1.flatbuffers.Builder(1);
        // Create modifications
        const modificationsArray = [];
        this.modifications.forEach((modification) => {
            const addressModificationVector = RestrictionMosaicModificationBuffer
                .createValueVector(builder, modification.value);
            RestrictionMosaicModificationBuffer.startRestrictionMosaicModificationBuffer(builder);
            RestrictionMosaicModificationBuffer.addModificationType(builder, modification.type);
            RestrictionMosaicModificationBuffer.addValue(builder, addressModificationVector);
            modificationsArray.push(RestrictionMosaicModificationBuffer.endRestrictionMosaicModificationBuffer(builder));
        });
        // Create vectors
        const signatureVector = AccountRestrictionsMosaicTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = AccountRestrictionsMosaicTransactionBuffer
            .createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = AccountRestrictionsMosaicTransactionBuffer
            .createDeadlineVector(builder, this.deadline);
        const feeVector = AccountRestrictionsMosaicTransactionBuffer
            .createFeeVector(builder, this.maxFee);
        const modificationVector = AccountRestrictionsMosaicTransactionBuffer
            .createModificationsVector(builder, modificationsArray);
        AccountRestrictionsMosaicTransactionBuffer.startAccountRestrictionsMosaicTransactionBuffer(builder);
        AccountRestrictionsMosaicTransactionBuffer.addSize(builder, 122 + (9 * this.modifications.length));
        AccountRestrictionsMosaicTransactionBuffer.addSignature(builder, signatureVector);
        AccountRestrictionsMosaicTransactionBuffer.addSigner(builder, signerVector);
        AccountRestrictionsMosaicTransactionBuffer.addVersion(builder, this.version);
        AccountRestrictionsMosaicTransactionBuffer.addType(builder, this.type);
        AccountRestrictionsMosaicTransactionBuffer.addFee(builder, feeVector);
        AccountRestrictionsMosaicTransactionBuffer.addDeadline(builder, deadlineVector);
        AccountRestrictionsMosaicTransactionBuffer.addRestrictionType(builder, this.restrictionType);
        AccountRestrictionsMosaicTransactionBuffer.addModificationCount(builder, this.modifications.length);
        AccountRestrictionsMosaicTransactionBuffer.addModifications(builder, modificationVector);
        // Calculate size
        const codedAccountRestrictionsMosaic = AccountRestrictionsMosaicTransactionBuffer.endAccountRestrictionsMosaicTransactionBuffer(builder);
        builder.finish(codedAccountRestrictionsMosaic);
        const bytes = builder.asUint8Array();
        return new AccountRestrictionsMosaicTransaction(bytes);
    }
}
exports.Builder = Builder;
