// nem2-sdk/dist/src/infrastructure/builders/MosaicAddressRestrictionTransaction.js
"use strict";
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @module transactions/MosaicAddressRestrictionTransaction
 */
const TransactionType_1 = require("../../model/transaction/TransactionType");
const MosaicAddressRestrictionTransactionBuffer_1 = require("../buffers/MosaicAddressRestrictionTransactionBuffer");
const MosaicAddressRestrictionTransactionSchema_1 = require("../schemas/MosaicAddressRestrictionTransactionSchema");
const VerifiableTransaction_1 = require("./VerifiableTransaction");
const flatbuffers_1 = require("flatbuffers");
const format_1 = require("../../core/format");
const { MosaicAddressRestrictionTransactionBuffer, } = MosaicAddressRestrictionTransactionBuffer_1.default.Buffers;
class MosaicAddressRestrictionTransaction extends VerifiableTransaction_1.VerifiableTransaction {
    constructor(bytes, schema) {
        super(bytes, schema);
    }
}
exports.default = MosaicAddressRestrictionTransaction;
// tslint:disable-next-line: max-classes-per-file
class Builder {
    constructor() {
        this.maxFee = [0, 0];
        this.type = TransactionType_1.TransactionType.MOSAIC_ADDRESS_RESTRICTION;
    }
    addFee(maxFee) {
        this.maxFee = maxFee;
        return this;
    }
    addVersion(version) {
        this.version = version;
        return this;
    }
    addType(type) {
        this.type = type;
        return this;
    }
    addDeadline(deadline) {
        this.deadline = deadline;
        return this;
    }
    addRestrictionKey(restrictionKey) {
        this.restrictionKey = restrictionKey;
        return this;
    }
    addTargetAddress(targetAddress) {
        this.targetAddress = format_1.RawAddress.stringToAddress(targetAddress);
        return this;
    }
    addPreviousRestrictionValue(previousRestrictionValue) {
        this.previousRestrictionValue = previousRestrictionValue;
        return this;
    }
    addNewRestrictionValue(newRestrictionValue) {
        this.newRestrictionValue = newRestrictionValue;
        return this;
    }
    addMosaicId(mosaicId) {
        this.mosaicId = mosaicId;
        return this;
    }
    build() {
        const builder = new flatbuffers_1.flatbuffers.Builder(1);
        // Create vectors
        const signatureVector = MosaicAddressRestrictionTransactionBuffer
            .createSignatureVector(builder, Array(...Array(64)).map(Number.prototype.valueOf, 0));
        const signerVector = MosaicAddressRestrictionTransactionBuffer
            .createSignerVector(builder, Array(...Array(32)).map(Number.prototype.valueOf, 0));
        const deadlineVector = MosaicAddressRestrictionTransactionBuffer
            .createDeadlineVector(builder, this.deadline);
        const feeVector = MosaicAddressRestrictionTransactionBuffer
            .createFeeVector(builder, this.maxFee);
        const mosaicIdVector = MosaicAddressRestrictionTransactionBuffer
            .createMosaicIdVector(builder, this.mosaicId);
        const restrictionKeyVector = MosaicAddressRestrictionTransactionBuffer
            .createRestrictionKeyVector(builder, this.restrictionKey);
        const targetAddressVector = MosaicAddressRestrictionTransactionBuffer
            .createTargetAddressVector(builder, this.targetAddress);
        const previousRestrictionValueVector = MosaicAddressRestrictionTransactionBuffer
            .createPreviousRestrictionValueVector(builder, this.previousRestrictionValue);
        const newRestrictionValueVector = MosaicAddressRestrictionTransactionBuffer
            .createNewRestrictionValueVector(builder, this.newRestrictionValue);
        MosaicAddressRestrictionTransactionBuffer.startMosaicAddressRestrictionTransactionBuffer(builder);
        MosaicAddressRestrictionTransactionBuffer.addSize(builder, 169);
        MosaicAddressRestrictionTransactionBuffer.addSignature(builder, signatureVector);
        MosaicAddressRestrictionTransactionBuffer.addSigner(builder, signerVector);
        MosaicAddressRestrictionTransactionBuffer.addVersion(builder, this.version);
        MosaicAddressRestrictionTransactionBuffer.addType(builder, this.type);
        MosaicAddressRestrictionTransactionBuffer.addFee(builder, feeVector);
        MosaicAddressRestrictionTransactionBuffer.addDeadline(builder, deadlineVector);
        MosaicAddressRestrictionTransactionBuffer.addMosaicId(builder, mosaicIdVector);
        MosaicAddressRestrictionTransactionBuffer.addRestrictionKey(builder, restrictionKeyVector);
        MosaicAddressRestrictionTransactionBuffer.addTargetAddress(builder, targetAddressVector);
        MosaicAddressRestrictionTransactionBuffer.addPreviousRestrictionValue(builder, previousRestrictionValueVector);
        MosaicAddressRestrictionTransactionBuffer.addNewRestrictionValue(builder, newRestrictionValueVector);
        // Calculate size
        const codedMosaicAddressRestriction = MosaicAddressRestrictionTransactionBuffer.endMosaicAddressRestrictionTransactionBuffer(builder);
        builder.finish(codedMosaicAddressRestriction);
        const bytes = builder.asUint8Array();
        const schema = MosaicAddressRestrictionTransactionSchema_1.default;
        return new MosaicAddressRestrictionTransaction(bytes, schema);
    }
}
exports.Builder = Builder;
